velox/expression/Expr.cpp

```diff
@@ -556,16 +556,24 @@ void Expr::eval(
   // all the time. Therefore, we should delay loading lazy vectors until we
   // know the minimum subset of rows needed to be loaded.
   //
-  // Load fields multiply referenced by inputs unconditionally. It's hard to
-  // know the superset of rows the multiple inputs need to load.
-  //
   // If there is only one field, load it unconditionally. The very first IF,
   // AND or OR will have to load it anyway. Pre-loading enables peeling of
   // encodings at a higher level in the expression tree and avoids repeated
   // peeling and wrapping in the sub-nodes.
   //
+  // Also load fields referenced by shared sub expressions to ensure that if
+  // there is an encoding on the loaded vector, then it is always peeled before
+  // evaluating sub-expression. Otherwise, the first call to
+  // evaluateSharedSubexpr might pass rows before peeling and the next one pass
+  // rows after peeling.
+  //
+  // Finally, for non-null propagating expressions, load multiply referenced
```

Comment:
nit: I looked into what this is trying to convey and I think a better description can be:
"
Finally, for non-null propagating expressions, load multiply referenced inputs unconditionally as it is hard to keep track of the superset of rows that would end up being evaluated among all its children (and hence need to be loaded). This is because any of the children might have null propagating expressions that end up operating on a reduced set of rows. So, one sub tree might need only a subset, whereas other might need a different subset.
"